### PR TITLE
Added new verbs: powershell_core and powershell

### PR DESCRIPTION
**Pull request by ProjectSynchro**
Saw an older PR that died a few years ago so I figured I'd try.

The Powershell Core verb is based on the Latest LTS release that supports both 32 and 64 bit Windows.

I didn't include dotnet48 as it isn't explicitly required by PowerShell Core and is quite heavy to install, it doesn't provide everything you would be using it for anyways. If it's wanted it can trivially be added.

The powershell verb is based on wrapper code is hosted at: https://github.com/ProjectSynchro/powershell-wrapper-for-wine, this is a fork of https://github.com/PietJankbal/powershell-wrapper-for-wine, I contacted the maintainer but they didn't get back to me so I went ahead and forked, if things change the links to the powershell wrapper binaries and profile.ps1 can be changed.

Let me know if there is anything I should do formatting-wise, I kind of copied what was around and attempted to keep things in alphabetical order. I would stick `powershell_core` after `powershell`, but with `powershell` requiring `powershell_core` it is what it is.

### CODE REVIEW

File: src/winetricks [hunk 1]
```diff
@@ -12292,6 +12292,81 @@ load_pngfilt()
 
 #----------------------------------------------------------------
 
+w_metadata powershell_core apps \
+    title="PowerShell Core" \
+    publisher="Microsoft" \
+    year="2024" \
+    media="download" \
+    file1="PowerShell-7.2.19-win-x64.msi" \
+    file2="PowerShell-7.2.19-win-x86.msi"
```
Comment (austin987) [inline src/winetricks:12301]: Pleas swap these (x86 should be file1)

File: src/winetricks [hunk 2]
```diff
@@ -12292,6 +12292,81 @@ load_pngfilt()
...
+# Function to download and install PowerShell Core
+load_powershell_core()
+{
+    #w_package_unsupported_win32
+
+    # Download PowerShell Core 7.2.x MSI (Latest LTS Version to support win32)
+    if [ "${W_ARCH}" = "win64" ]; then
+        # https://github.com/PowerShell/PowerShell/releases/v7.2.19
+        w_download https://github.com/PowerShell/PowerShell/releases/download/v7.2.19/PowerShell-7.2.19-win-x64.msi 965e10fc77dbce6ef622f1c0fcddc03bebd2b3629f1a3253cc17a8ffc71a6522
+        msi="PowerShell-7.2.19-win-x64.msi"
```
Comment (austin987) [inline src/winetricks:12313]: You could simplify this a bit by using ${fileX} here.

File: src/winetricks [hunk 3]
```diff
@@ -12292,6 +12292,81 @@ load_pngfilt()
...
+    elif [ "${W_ARCH}" = "win32" ]; then
+        # https://github.com/PowerShell/PowerShell/releases/v7.2.19
+        w_download https://github.com/PowerShell/PowerShell/releases/download/v7.2.19/PowerShell-7.2.19-win-x86.msi cbf214d41bc8be4b1ce5ba566dbfee5f335971c48231825e90af7954356ce797
+        msi="PowerShell-7.2.19-win-x86.msi"
```
Comment (austin987) [inline src/winetricks:12317]: Same.

File: src/winetricks [hunk 4]
```diff
@@ -12292,6 +12292,81 @@ load_pngfilt()
...
+
+load_powershell()
+{
+    w_do_call powershell_core
+
+    # Download PowerShell Wrapper profile.ps1
+    w_download https://github.com/ProjectSynchro/powershell-wrapper-for-wine/releases/download/v2.0.1/profile.ps1 85c7d4bc526a0b427cb8fbc77cff8a43b0475b2bf7397889cbe1ab224d1579a1
+
+    #Download PowerShell Wrapper 32bit exe
+        w_download https://github.com/ProjectSynchro/powershell-wrapper-for-wine/releases/download/v2.0.1/powershell32.exe e0c4192877ba2bce7830dc83745bfa54567e2a67c4e84b56953404d3ef0bca30
```
Comment (austin987) [inline src/winetricks:12346]: Indentation is off here.

File: src/winetricks [hunk 5]
```diff
@@ -12292,6 +12292,81 @@ load_pngfilt()
...
+    file3="profile.ps1"
+
+load_powershell()
+{
+    w_do_call powershell_core
+
+    # Download PowerShell Wrapper profile.ps1
+    w_download https://github.com/ProjectSynchro/powershell-wrapper-for-wine/releases/download/v2.0.1/profile.ps1 85c7d4bc526a0b427cb8fbc77cff8a43b0475b2bf7397889cbe1ab224d1579a1
+
+    #Download PowerShell Wrapper 32bit exe
```
Comment (austin987) [inline src/winetricks:12345]: Missing a space

**Review by austin987 — CHANGES_REQUESTED** [review]

File: src/winetricks (hunk 1 above)
Comment (ProjectSynchro) [reply src/winetricks]: Fixed by 509afca8a64d185256e28d9efd9e5e87bb71a6e3

File: src/winetricks (hunk 2 above)
Comment (ProjectSynchro) [reply src/winetricks]: Fixed by 06fca563298ecfca1a77cfb6079c2e7668f5fbd4 and 82c6052d9b778457ad0391453ad9b0ce0de04fbe

File: src/winetricks (hunk 3 above)
Comment (ProjectSynchro) [reply src/winetricks]: Fixed by 06fca563298ecfca1a77cfb6079c2e7668f5fbd4 and 82c6052d9b778457ad0391453ad9b0ce0de04fbe

File: src/winetricks (hunk 4 above)
Comment (ProjectSynchro) [reply src/winetricks]: Fixed by 509afca8a64d185256e28d9efd9e5e87bb71a6e3 and

File: src/winetricks (hunk 5 above)
Comment (ProjectSynchro) [reply src/winetricks]: Fixed by 509afca8a64d185256e28d9efd9e5e87bb71a6e3

**Comment by ProjectSynchro** [comment]
I tried to derive as much as possible from the metadata (including version from file1), let me know if you'd like any more changes.

**Comment by mactan-sc** [comment]
looking forward to having a verb for this, the star citizen game launcher is unable to perform two critical functions without powershell

**Comment by ProjectSynchro** [comment]
@austin987 
Is there anything else that you'd like done to get this merged?

AFAIK I have actioned everything you had suggested.

Cheers.

File: src/winetricks [hunk 6]
```diff
@@ -12292,6 +12292,97 @@ load_pngfilt()
...
 #----------------------------------------------------------------
 
+w_metadata powershell_core apps \
+    title="PowerShell Core" \
+    publisher="Microsoft" \
+    year="2024" \
+    media="download" \
+    file1="PowerShell-7.2.19-win-x86.msi" \
+    file2="PowerShell-7.2.19-win-x64.msi"
+
```
Comment (austin987) [inline src/winetricks:12302]: There's trailing whitespace here.

File: src/winetricks [hunk 7]
```diff
@@ -12292,6 +12292,97 @@ load_pngfilt()
...
+    _load_powershell_core_regex='[0-9]+\.[0-9]+\.[0-9]+'
+
+    if echo "${file1}" | grep -E -o "${_load_powershell_core_regex}" >/dev/null; then
+        _load_powershell_core_version=$(echo "${file1}" | grep -E -o "${_load_powershell_core_regex}")
+    fi
+
+    # Download PowerShell Core 7.2.x MSI (Latest LTS Version to support win32)
+    if [ "${W_ARCH}" = "win64" ]; then
+        # https://github.com/PowerShell/PowerShell/releases/v7.2.19
+        w_download "https://github.com/PowerShell/PowerShell/releases/download/v${_load_powershell_core_version}/${file2}" 965e10fc77dbce6ef622f1c0fcddc03bebd2b3629f1a3253cc17a8ffc71a6522
```
Comment (austin987) [inline src/winetricks:12321]: Please don't use variables with w_download URLs, it breaks src/linkcheck.sh.

BTW, if you run ./tests/shell-checks before submitting, you'll get an error letting you know that.

**Review by austin987 — CHANGES_REQUESTED** [review]

File: src/winetricks (hunk 6 above)
Comment (ProjectSynchro) [reply src/winetricks]: Fixed as of https://github.com/Winetricks/winetricks/pull/2211/commits/1d323f605c4039dd2a120530c2d22926b038a52c

File: src/winetricks (hunk 7 above)
Comment (ProjectSynchro) [reply src/winetricks]: Good to know, I have fixed this as of https://github.com/Winetricks/winetricks/pull/2211/commits/b51fbac1d52db177162b5a87425d1c43da1dff7a

./tests/shell-checks appears to be happy after running a test.

**Comment by ProjectSynchro** [comment]
I cleaned up the commit history.

Let me know if there are any additional changes needed, from what I can see I think it should be good to go, unless there's another linter I haven't found to run :smile:

File: src/winetricks [hunk 8]
```diff
@@ -12413,6 +12413,84 @@ load_pngfilt()
...
+load_powershell_core()
+{
+    # Uncomment below and remove win32 download elif and file when PowerShell Core v7.2 LTS goes EOL
+    #w_package_unsupported_win32
+
+    # Download PowerShell Core 7.2.x MSI (Latest LTS Version to support win32)
+    # https://github.com/PowerShell/PowerShell/releases/v7.2.21
+    if [ "${W_ARCH}" = "win64" ]; then
+        w_download "https://github.com/PowerShell/PowerShell/releases/download/v7.2.21/PowerShell-7.2.21-win-x64.msi" 407640b11c89d66ec7892229e68b1d74b26f0e820b52da268c67fd166c2b46ad
+        msi="${file2}"
```
Comment (austin987) [inline src/winetricks:12434]: Almost there :).

This causes a shellcheck error (see the github actions).

Shellcheck doesn't know that the metadata is sourced before the function is called. You can either disable the warning, or remove the variables and hardcode it, up to you.

File: src/winetricks [hunk 9]
```diff
@@ -12413,6 +12413,84 @@ load_pngfilt()
 
 #----------------------------------------------------------------
 
+w_metadata powershell_core apps \
```
Comment (austin987) [inline src/winetricks:12416]: Sorry I missed this before, but it's probably better to use 'dlls' here.

File: src/winetricks [hunk 10]
```diff
@@ -12413,6 +12413,84 @@ load_pngfilt()
...
+    # Change directory to the cache directory where the MSI file is downloaded
+    w_try_cd "${W_CACHE}/${W_PACKAGE}"
+
+    # Install PowerShell Core using Wine's msiexec
+    w_try "${WINE}" msiexec ${W_OPT_UNATTENDED:+/quiet} /i "${msi}" ENABLE_PSREMOTING=1 REGISTER_MANIFEST=1
+}
+
+#----------------------------------------------------------------
+
+w_metadata powershell apps \
```
Comment (austin987) [inline src/winetricks:12449]: Same here.

Comment (ProjectSynchro) [reply src/winetricks]: fixed with https://github.com/Winetricks/winetricks/pull/2211/commits/5e61e99ecc5ec88d18c1b7ed5b6e2ba46bcd2c2c

File: src/winetricks (hunk 8 above)
Comment (ProjectSynchro) [reply src/winetricks:12436]: fixed with https://github.com/Winetricks/winetricks/pull/2211/commits/768e913101d419be762a93c142e2814808179d8d

File: src/winetricks (hunk 9 above)
Comment (ProjectSynchro) [reply src/winetricks]: fixed with https://github.com/Winetricks/winetricks/pull/2211/commits/768e913101d419be762a93c142e2814808179d8d

**Comment by ProjectSynchro** [comment]
Decided to just go with the "silence shellcheck" route, hopefully that's okay. :smile: 

Ran the same script your CI does to verify scripts and that seems to have passed.

**Comment by austin987** [comment]
> Decided to just go with the "silence shellcheck" route, hopefully that's okay. 😄

Yep, that's fine.

> Ran the same script your CI does to verify scripts and that seems to have passed.

Are you sure :)? It fails for GitHub Actions with 0.8.0 (and with 0.10.0 for me locally):
```
In src/winetricks line 12489:
        w_try cp "${file3}" "${W_PROGRAMW6432_UNIX}/PowerShell/7/${file3}"
                  ^------^ SC2154 (warning): file3 is referenced but not assigned (did you mean 'file'?).
```
What version of shellcheck did this work with?

AFAIK Shellcheck directives have to be on the line above the offending line (or at the top of the file, or before the function name), see https://www.shellcheck.net/wiki/Directive. That said, to be fair, I'm not sure why it worked for the other cases.

The safest way to handle it is to put the directive directly above the affected lines (that way if any future variables are added to these functions, they're not affected).

**Comment by ProjectSynchro** [comment]
> Are you sure :)? It fails for GitHub Actions with 0.8.0 (and with 0.10.0 for me locally):
> 
> ```
> In src/winetricks line 12489:
>         w_try cp "${file3}" "${W_PROGRAMW6432_UNIX}/PowerShell/7/${file3}"
>                   ^------^ SC2154 (warning): file3 is referenced but not assigned (did you mean 'file'?).
> ```
> 
> What version of shellcheck did this work with?
> 

This worked with 0.10.0, but after testing this again it looks like the shell-checks script fails silently if bashate is not installed. I'm not sure if this is intentional or not but I'm guessing not :)

Installing it has the script running correctly, and I was able to repro the issue you found.

> AFAIK Shellcheck directives have to be on the line above the offending line (or at the top of the file, or before the function name), see https://www.shellcheck.net/wiki/Directive. That said, to be fair, I'm not sure why it worked for the other cases.
> 
> The safest way to handle it is to put the directive directly above the affected lines (that way if any future variables are added to these functions, they're not affected).

Agreed, I suspect since the scripts are interpreted, it disables the checks in whatever context the directive is located. I stuck them above each of the effected lines

**Review by austin987 — APPROVED** [review]

**Comment by austin987** [comment]
Thanks (for the PR, and your patience :) )!

**Pull request merged**